### ISSUE TITLE
How do I handle multi-select boxes

### ISSUE THREAD

**Issue by JackWillDavis** [mikehaertl/php-pdftk#148]
Hi,

I've seen some similar questions but none which help in my particular use-case. How do I handle groups of checkboxes which have the same name but multiple possible options for the user using the PHP Library?

For example, say I have a list of animals in a PDF and the user ticks which ones they own:

```
What pets do you have?
* Dog (name: animals_select, export value: dog)
* Cat (name: animals_select, export value: cat)
* House Pig (name: animals_select, export value: house_pig)
```
I was hoping I could pass multiple values through to the library like so:

```
$pdf = new PDF("path/to/template.pdf");
$pdf->fillForm(array(
   'user_name' => 'John Smith',
   'animals_select' => array(
      'dog',
      'house_pig
   )
));
```

The ideal outcome being when we received the PDF, it would have the **Dog** and **House Pig** checkboxes ticked in that group.

However, from looking at the `XfdfFile.php` file, it looks like I can only pass a single value for every form field on the page. So, I edited the file a little bit to accept arrays (it was previously failing on the sanitizing if it was an array) but although it reaches the `writeFields()` function properly formatted, the boxes still don't get ticked on the final PDF.

Please could you point me in the right direction or let me know if this is something which isn't possible currently?

For reference, here's what I changed on the `writeFields()` function to allow multiple values for the same key.

```
protected function writeFields($fp, $fields, $recKey = false)
    {
        foreach ($fields as $key => $value) {
            if($recKey){
              fwrite($fp, "<field name=\"$recKey\">\n<value>$value</value>\n</field>\n");
            } else if (is_array($value)) {
              fwrite($fp, "<field name=\"$key\">\n");
              $this->writeFields($fp, $value, $key);
              fwrite($fp, "</field>\n");
            } else {
              fwrite($fp, "<field name=\"$key\">\n<value>$value</value>\n</field>\n");
            }
        }
    }
```

I feel like it *should* be working but I might be missing an easier way for the library to handle it, or I'm misunderstanding how multi-select boxes are handled in PDFs. Any advice on how to achieve this with your library would be greatly appreciated!

**Comment by mikehaertl** [comment]
This should already work. In your example you use different form names in the PDF and the PHP array: `animals_select` vs. `animals`.

**Comment by mikehaertl** [comment]
Sorry, maybe I was wrong. To be honest I'm not sure. I thought I already had such a case in one of my projects, but can't find any. So if you could somehow find out, how the XFD must look in this case, we can probably add a fix.

**Comment by JackWillDavis** [comment]
Ahh that was a typo from me as I've changed the keys for this question. I'm 100% sure I'm using the same keys in the PDF file and the PHP.

**Comment by mikehaertl** [comment]
Yeah, sorry, see my post above. It would help if you could just open the XFDF, edit it, and pass it to `pdftk fill_form`. Maybe you can figure out the correct format.

**Comment by JackWillDavis** [comment]
OK thanks, to be honest I'm a little relieved I haven't spent an hour or two trying to get it work and there was an easier way.

Alright so can I confirm the theory then? If I amend the `XfdfFile.php` to write multiple records like:

```
"<field name="animals_select">\n<value>dog</value>\n</field>\n"
"<field name="animals_select">\n<value>house_pig</value>\n</field>\n"
```

It should check those relevant boxes on the PDF?

**Update**

I've just seen your response, thanks for getting back to me. I couldn't find that function in your repo so is that in core PDFTK?

**Comment by mikehaertl** [comment]
What I meant: Ignore my library here, just create a template XFDF file (you can use this library for this: https://github.com/mikehaertl/php-pdftk#create-a-xfdffdf-file-from-a-php-array).

Then use `pdftk` on the command line to fill your form. It also accepts XFDF files (at least I think so). Then modify the XFDF file until it works. I'm sorry I can't help more here. Just don't have the time for it. But if we know what the correct XFDF XML format for multiple checkboxes is, we should be able to fix the library.

**Comment by mikehaertl** [comment]
One more idea: Did you try this structure:

```xml
<field name="x">
   <value>a</value>
   <value>b</value>
</field>
```

**Comment by JackWillDavis** [comment]
Hey thanks a lot for that, I've just tried that format actually. I can see it's being written to the file correctly (as below) but it's not ticking them unfortunately. I'm just going to look up, I have a feeling it needs a `type` declaring on `<field>` but I'm not sure.

```
<field name="animals_select">
   <value>dog</value>
   <value>house_pig</value>
</field>
```

**Comment by mikehaertl** [comment]
Not too sure about the type. I found the XFDF specs here, and the format is quite simple: https://www.immagic.com/eLibrary/ARCHIVES/TECH/ADOBE/A070914X.pdf

But there's nothing about multiple field values.

Another idea: Maybe you find some PDF viewer where you can fill out the form and then save the form data as XFDF.

**Comment by JackWillDavis** [comment]
Thanks for your help with this. After a lot of digging, it's looking like if you give multiple checkboxes in a PDF document the same name and different export values, they're treated as a radio button where only one can be selected at once. Sounds a bit mad but after a little testing it looks like it is actually acting that way in Acrobat DC too.

I might be wrong but I found a few discussions where people seem to be saying the same thing [here](https://answers.acrobatusers.com/How-do-I-format-check-boxes-so-only-one-will-be-checked-at-a-time-q3382.aspx) and [here](https://forums.adobe.com/thread/923348)

So, I guess that's why there's nothing in the documentation about it. We're going to get around this by just giving each checkbox a unique name and default export value.

Thanks for taking the time out though, very much appreciated and please let me know if I'm wrong about it as I'd be up for working on it and putting a PR across.

**Comment by mikehaertl** [comment]
Now as you say it, it makes perfect sense: PDF forms are not like HTML forms, so there's not neccessarily the same thing like a checkbox list with a single name in PDF. It's simply not supported.

I also remember now, that in my project each checkbox had it's own name. Guess you agree, that I can close this.

**Comment by maxguru** [comment]
List boxes can have multiple values selected simultaneously.

**Comment by maxguru** [comment]
@mikehaertl, filling multi-select list boxes doesn't work right now.

The following code yields the following result:
```
$data = ["multi-select list box" => ["item1", "item2"]];
$xfdf = new \mikehaertl\pdftk\XfdfFile($data);
$xfdf->saveAs('/tmp/data.xfdf');
print(file_get_contents('/tmp/data.xfdf'));
```
```
<?xml version="1.0" encoding="UTF-8"?>
<xfdf xmlns="http://ns.adobe.com/xfdf/" xml:space="preserve">
<fields><field name="multi-select list box">
<value></value>
</field>
</fields>
</xfdf>
```

I can submit a pull request to fix this issue, if you are interested. Let me know.

**Comment by mikehaertl** [comment]
@maxguru I'll have a look. Could you also post the expected XML? Just to make sure we talk about the same expectations.

**Comment by mikehaertl** [comment]
Reading the [XFDF specs](https://www.immagic.com/eLibrary/ARCHIVES/TECH/ADOBE/A070914X.pdf) I think that a field *can* have multiple `<value>` tags:

> field
> The field element is a child of the fields and field elements. The field element corresponds to a
> form field. 
>
> Content model
> ( field* | value* | (value? & value-richtext?) )

@maxguru I've created a PR. Can you help testing?

**Comment by maxguru** [comment]
Thanks for the quick response!

The new code outputs the following:
```
<?xml version="1.0" encoding="UTF-8"?>
<xfdf xmlns="http://ns.adobe.com/xfdf/" xml:space="preserve">
<fields>
<field name="multi-select list box">
<value>item1</value>
<value>item2</value>
</field>
</fields>
</xfdf>
```

However, for whatever reason pdftk outputs a PDF file which has the last item selected, not both of them.

I use the following pdftk command:
```
pdftk input.pdf fill_form test.xfdf output test.pdf
```

```
$ pdftk --version
pdftk port to java 3.2.2 a Handy Tool for Manipulating PDF Documents
Copyright (c) 2017-2018 Marc Vinyals - https://gitlab.com/pdftk-java/pdftk
Copyright (c) 2003-2013 Steward and Lee, LLC.
pdftk includes a modified version of the iText library.
Copyright (c) 1999-2009 Bruno Lowagie, Paulo Soares, et al.
This is free software; see the source code for copying conditions. There is
NO warranty, not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by mikehaertl** [comment]
I can't tell you, why that is. Maybe a limitation/bug with pdftk?

Have you tried loading the XFDF with some original Adobe tools? Does this work at all?

It would also be interesting to see the XFDF generated by Adobe tools.

**Comment by maxguru** [comment]
Let me get back to you once I find out more.

**Comment by maxguru** [comment]
Alright, I've tried a bunch of tools and most of them have some sort of issue with multi-select either with XFDF or XFDF and FDF export/import.

Acrobat does import the multi-select XFDF/FDF correctly and exports XFDF semi-correctly:
```
...
><field name="List Box Multiselect"
><value
>item2</value
><value
>item3</value
></field
></fields
...
```

So, your pull request code is generating a usable XFDF format that works with Acrobat but pdftk will need to be fixed.

**Comment by maxguru** [comment]
By the way, I've opened an [issue](https://gitlab.com/pdftk-java/pdftk/-/issues/101) with pdftk-java over at GitLab.

**Comment by mikehaertl** [comment]
Alright, thanks for the update. I've merged the changes and released 0.10.2. I've removed the notes from the README, though, as forms still can't be filled with pdftk this way.

Just wanted to suggest to open an issue with pdftk-java but I see you already did: https://gitlab.com/pdftk-java/pdftk/-/issues/101

So let's see if they fix this. Just a note: It would probably help them if you could create a very simple PDF + XFDF file with a multiselect field to demonstrate the issue.
